Modify rule S5000: LaYC format (#1909)
SonarSource/rspec: rules/S5000/cfamily/rule.adoc

> The function ``++memcmp++`` only returns meaningful results for objects of trivially copyable types that have no padding. This includes scalar types, arrays, and trivially copyable classes.

Reviewer: ```suggestion
The function ``++memcmp++`` only returns meaningful results for objects of trivially copyable types without padding. This includes scalar types, arrays, and trivially copyable classes.
```

== Why is this an issue?

> `memcmp` compares the raw memory representation of two objects (what the standard calls their _object representation_). For objects that are not trivially copyable or that contain padding, two objects storing identical values could have different raw memory representations, so the result of `memcmp` is not meaningful.

Reviewer: ```suggestion
`memcmp` compares the raw memory representation of two objects (what the standard calls their _object representation_). When objects are not trivially copyable or contain padding, they could have different raw memory representations even though they store identical values. So the result of `memcmp` is not meaningful.
```

A class type is trivially copyable if:

> * One or more of the following methods is trivial and the rest are deleted: copy /constructor, move constructor, copy assignment operator, and move assignment operator,

Reviewer: What is a trivial copy constructor? A trivial move constructor? A trivial destructor? etc...
I think this should also be explained, don't you think?

Author: Added explanations.

* It has a trivial, non-deleted destructor.

Padding is the addition of extra bits to a storage unit to address memory alignment. If the type contains padding, some of its bits might be non-representative.

== How to fix it

=== Code examples

==== Noncompliant code example

[source,cpp,diff-id=1,diff-type=noncompliant]
----

> class Shape { // Trivially copyable, but will contain padding after the bool on most architectures

Reviewer: Do you think we can make the first example more C oriented?

Author: Done

[source,c,diff-id=1,diff-type=noncompliant]
----
struct Shape { // Trivially copyable, but will contain padding after the bool on most architectures
  bool visible;
  int x;
  int y;
};

bool isSame(Shape *s1, Shape *s2)
{
    return memcmp(s1, s2, sizeof(Shape)) == 0; // Noncompliant
}
----

[source,cpp,diff-id=2,diff-type=noncompliant]
----
class Resource { // Not trivially copyable
public:
  Ptr* ptr;
  ~Resource();
};

bool isSame(Resource *r1, Resource *r2)
{
    return memcmp(r1, r2, sizeof(Resource)) == 0; // Noncompliant
}
----

==== Compliant solution

[source,c,diff-id=1,diff-type=compliant]
----
struct Shape { // Trivially copyable, but will contain padding after the bool on most architectures
  bool visible;
  int x;
  int y;
};

bool isSame(Shape *s1, Shape *s2)
{
  return s1->visible == s2->visible && s1->x == s2->x && s1->y == s2->y;
}
----

[source,cpp,diff-id=2,diff-type=compliant]
----
class Resource { // Not trivially copyable
public:
  Ptr* ptr;
  ~Resource();
};

bool operator==(Resource const &r1, Resource const &r2) {
  return r1.ptr == r2.ptr;
}

bool isSame(Resource *r1, Resource *r2)
{
    return (*r1) == (*r2);
}
----

== Resources

> https://en.cppreference.com/w/cpp/language/classes#Trivially_copyable_class[Detailed definition of what a trivially copyable class is].

Reviewer: ```suggestion
https://en.cppreference.com/w/cpp/language/classes#Trivially_copyable_class[Definition of a trivially copyable class].
```
